# **DATA PREPROCESSING**

## Importing Required Libraries

In [1]:
# DATA HANDLING
import numpy as np
import pandas as pd

# DATA VISUALIZATION
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

## Loading the Dataset

In [2]:
# ===Load Data===
pd.set_option('display.max_columns', None)
data=pd.read_excel("INX_Future_Inc_Employee_Performance_CDS_Project2_Data_V1.8.xls")
data.to_csv('data/raw/INX_Future_Inc_Employee_Performance.csv', index=False)
# ===Checking first five rows of dataset===
data.head()

,EmpNumber,Age,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,EmpHourlyRate,EmpJobInvolvement,EmpJobLevel,EmpJobSatisfaction,NumCompaniesWorked,OverTime,EmpLastSalaryHikePercent,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,Attrition,PerformanceRating
0,E1001000,32,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,10,3,4,55,3,2,4,1,No,12,4,10,2,2,10,7,0,8,No,3
1,E1001006,47,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,14,4,4,42,3,2,1,2,No,12,4,20,2,3,7,7,1,7,No,3
2,E1001007,40,Male,Life Sciences,Married,Sales,Sales Executive,Travel_Frequently,5,4,4,48,2,3,1,5,Yes,21,3,20,2,3,18,13,1,12,No,4
3,E1001009,41,Male,Human Resources,Divorced,Human Resources,Manager,Travel_Rarely,10,4,2,73,2,5,4,3,No,15,2,23,2,2,21,6,12,6,No,3
4,E1001010,60,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,16,4,1,84,3,2,1,8,No,14,4,10,1,3,2,2,2,2,No,3


In [3]:
# ===Checking the last five rows of dataset===
data.tail()

,EmpNumber,Age,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,EmpHourlyRate,EmpJobInvolvement,EmpJobLevel,EmpJobSatisfaction,NumCompaniesWorked,OverTime,EmpLastSalaryHikePercent,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,Attrition,PerformanceRating
1195,E100992,27,Female,Medical,Divorced,Sales,Sales Executive,Travel_Frequently,3,1,4,71,4,2,4,1,Yes,20,2,6,3,3,6,5,0,4,No,4
1196,E100993,37,Male,Life Sciences,Single,Development,Senior Developer,Travel_Rarely,10,2,4,80,4,1,4,3,No,17,1,4,2,3,1,0,0,0,No,3
1197,E100994,50,Male,Medical,Married,Development,Senior Developer,Travel_Rarely,28,1,4,74,4,1,3,1,Yes,11,3,20,3,3,20,8,3,8,No,3
1198,E100995,34,Female,Medical,Single,Data Science,Data Scientist,Travel_Rarely,9,3,4,46,2,3,2,1,No,14,2,9,3,4,8,7,7,7,No,3
1199,E100998,24,Female,Life Sciences,Single,Sales,Sales Executive,Travel_Rarely,3,2,1,65,3,2,3,9,No,14,1,4,3,3,2,2,2,0,Yes,2


In [4]:
# === Checking the no of rows and columns ===
print('We have {} rows and {} columns'.format(data.shape[0],data.shape[1]))

We have 1200 rows and 28 columns


In [5]:
# === Checking the info of dataset ===
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 28 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   EmpNumber                     1200 non-null   object
 1   Age                           1200 non-null   int64 
 2   Gender                        1200 non-null   object
 3   EducationBackground           1200 non-null   object
 4   MaritalStatus                 1200 non-null   object
 5   EmpDepartment                 1200 non-null   object
 6   EmpJobRole                    1200 non-null   object
 7   BusinessTravelFrequency       1200 non-null   object
 8   DistanceFromHome              1200 non-null   int64 
 9   EmpEducationLevel             1200 non-null   int64 
 10  EmpEnvironmentSatisfaction    1200 non-null   int64 
 11  EmpHourlyRate                 1200 non-null   int64 
 12  EmpJobInvolvement             1200 non-null   int64 
 13  EmpJobLevel       

The dataset has **1200 rows and 28 columns**, with **19 numerical and 9 categorical features**. There is **no missing values**.

In [6]:
# === To view the summary stats of numerical columns ===
data.describe()

,Age,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,EmpHourlyRate,EmpJobInvolvement,EmpJobLevel,EmpJobSatisfaction,NumCompaniesWorked,EmpLastSalaryHikePercent,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,PerformanceRating
count,1200.000000,1200.000000,1200.00000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000
mean,36.918333,9.165833,2.89250,2.715833,65.981667,2.731667,2.067500,2.732500,2.665000,15.222500,2.725000,11.330000,2.785833,2.744167,7.077500,4.291667,2.194167,4.105000,2.948333
std,9.087289,8.176636,1.04412,1.090599,20.211302,0.707164,1.107836,1.100888,2.469384,3.625918,1.075642,7.797228,1.263446,0.699374,6.236899,3.613744,3.221560,3.541576,0.518866
min,18.000000,1.000000,1.00000,1.000000,30.000000,1.000000,1.000000,1.000000,0.000000,11.000000,1.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,2.000000
25%,30.000000,2.000000,2.00000,2.000000,48.000000,2.000000,1.000000,2.000000,1.000000,12.000000,2.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000,3.000000
50%,36.000000,7.000000,3.00000,3.000000,66.000000,3.000000,2.000000,3.000000,2.000000,14.000000,3.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000,3.000000
75%,43.000000,14.000000,4.00000,4.000000,83.000000,3.000000,3.000000,4.000000,4.000000,18.000000,4.000000,15.000000,3.000000,3.000000,10.000000,7.000000,3.000000,7.000000,3.000000
max,60.000000,29.000000,5.00000,4.000000,100.000000,4.000000,5.000000,4.000000,9.000000,25.000000,4.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000,4.000000


In [7]:
# === To View the categorical columns ===
data.describe(include='O')

,EmpNumber,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,OverTime,Attrition
count,1200,1200,1200,1200,1200,1200,1200,1200,1200
unique,1200,2,6,3,6,19,3,2,2
top,E100967,Male,Life Sciences,Married,Sales,Sales Executive,Travel_Rarely,No,No
freq,1,725,492,548,373,270,846,847,1022


In [8]:
# ============Let's have a look at how many labels each variable has==========
for col in data.columns:
  print(col,':', len(data[col].unique()),'labels')

EmpNumber : 1200 labels
Age : 43 labels
Gender : 2 labels
EducationBackground : 6 labels
MaritalStatus : 3 labels
EmpDepartment : 6 labels
EmpJobRole : 19 labels
BusinessTravelFrequency : 3 labels
DistanceFromHome : 29 labels
EmpEducationLevel : 5 labels
EmpEnvironmentSatisfaction : 4 labels
EmpHourlyRate : 71 labels
EmpJobInvolvement : 4 labels
EmpJobLevel : 5 labels
EmpJobSatisfaction : 4 labels
NumCompaniesWorked : 10 labels
OverTime : 2 labels
EmpLastSalaryHikePercent : 15 labels
EmpRelationshipSatisfaction : 4 labels
TotalWorkExperienceInYears : 40 labels
TrainingTimesLastYear : 7 labels
EmpWorkLifeBalance : 4 labels
ExperienceYearsAtThisCompany : 37 labels
ExperienceYearsInCurrentRole : 19 labels
YearsSinceLastPromotion : 16 labels
YearsWithCurrManager : 18 labels
Attrition : 2 labels
PerformanceRating : 3 labels


In [9]:
#============ Getting catagorial and numerical Features ================

numerical_feature= [feature for feature in data.columns if data[feature].dtype!='O']
catagorical_feature=[feature for feature in data.columns if data[feature].dtype=='O']

print('\nWe have {} numerical feature:{}'.format(len(numerical_feature),numerical_feature))

print('\n\nWe have {} catagorical feature:{}'.format(len(catagorical_feature),catagorical_feature))


We have 19 numerical feature:['Age', 'DistanceFromHome', 'EmpEducationLevel', 'EmpEnvironmentSatisfaction', 'EmpHourlyRate', 'EmpJobInvolvement', 'EmpJobLevel', 'EmpJobSatisfaction', 'NumCompaniesWorked', 'EmpLastSalaryHikePercent', 'EmpRelationshipSatisfaction', 'TotalWorkExperienceInYears', 'TrainingTimesLastYear', 'EmpWorkLifeBalance', 'ExperienceYearsAtThisCompany', 'ExperienceYearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager', 'PerformanceRating']


We have 9 catagorical feature:['EmpNumber', 'Gender', 'EducationBackground', 'MaritalStatus', 'EmpDepartment', 'EmpJobRole', 'BusinessTravelFrequency', 'OverTime', 'Attrition']


## 1. Department-wise Performance

In [10]:
department_performance = data.groupby("EmpDepartment")["PerformanceRating"].mean().sort_values(ascending=False)
print("\nDepartment-wise Performance Ratings:\n", department_performance)


Department-wise Performance Ratings:
 EmpDepartment
Development               3.085873
Data Science              3.050000
Human Resources           2.925926
Research & Development    2.921283
Sales                     2.860590
Finance                   2.775510
Name: PerformanceRating, dtype: float64


#### Insights
##### **Department-wise Performance**
**1. Top Performing Departments:**
* Development (3.08)
* Data Science (3.05)
  
**2. Lowest Performing Departments:**
* Sales (2.86)
* Finance (2.77)

In [11]:
# ==== Finding a missing values ====
data.isnull().sum()

EmpNumber                       0
Age                             0
Gender                          0
EducationBackground             0
MaritalStatus                   0
EmpDepartment                   0
EmpJobRole                      0
BusinessTravelFrequency         0
DistanceFromHome                0
EmpEducationLevel               0
EmpEnvironmentSatisfaction      0
EmpHourlyRate                   0
EmpJobInvolvement               0
EmpJobLevel                     0
EmpJobSatisfaction              0
NumCompaniesWorked              0
OverTime                        0
EmpLastSalaryHikePercent        0
EmpRelationshipSatisfaction     0
TotalWorkExperienceInYears      0
TrainingTimesLastYear           0
EmpWorkLifeBalance              0
ExperienceYearsAtThisCompany    0
ExperienceYearsInCurrentRole    0
YearsSinceLastPromotion         0
YearsWithCurrManager            0
Attrition                       0
PerformanceRating               0
dtype: int64

In [12]:
# ==== Check duplicate values ====
# ==== Total number of rows ====
total_rows = len(data)

# ==== Count duplicate rows ====
duplicate_count = data.duplicated().sum()

# ==== Percentage of duplicates ====
duplicate_percentage = (duplicate_count / total_rows) * 100

print(f"Total Rows: {total_rows}")
print(f"Duplicate Rows: {duplicate_count}")
print(f"Percentage of Duplicates: {duplicate_percentage:.2f}%")

Total Rows: 1200
Duplicate Rows: 0
Percentage of Duplicates: 0.00%


Out of 1200 total records, there is **no duplicate entries** were identified.

In [13]:
#  =============== Checking the value_counts =======================
cat_data = data[['Gender', 'EducationBackground', 'MaritalStatus','EmpDepartment', 'EmpJobRole',
       'BusinessTravelFrequency', 'OverTime', 'Attrition']]

for i in cat_data.columns:
    print(data[i].value_counts())
    print('============================')
    print("\n")

Gender
Male      725
Female    475
Name: count, dtype: int64


EducationBackground
Life Sciences       492
Medical             384
Marketing           137
Technical Degree    100
Other                66
Human Resources      21
Name: count, dtype: int64


MaritalStatus
Married     548
Single      384
Divorced    268
Name: count, dtype: int64


EmpDepartment
Sales                     373
Development               361
Research & Development    343
Human Resources            54
Finance                    49
Data Science               20
Name: count, dtype: int64


EmpJobRole
Sales Executive              270
Developer                    236
Manager R&D                   94
Research Scientist            77
Sales Representative          69
Laboratory Technician         64
Senior Developer              52
Manager                       51
Finance Manager               49
Human Resources               45
Technical Lead                38
Manufacturing Director        33
Healthcare Representative 

In [14]:
#========== Target Variable Profiling Before Exploratory Data Analysis ===========
target = "PerformanceRating"

print("Counts:\n", data[target].value_counts())
print("\nPercentages:\n", data[target].value_counts(normalize=True)*100)
print('-'*80)
print("\nUnique classes:", data[target].nunique())

Counts:
 PerformanceRating
3    874
2    194
4    132
Name: count, dtype: int64

Percentages:
 PerformanceRating
3    72.833333
2    16.166667
4    11.000000
Name: proportion, dtype: float64
--------------------------------------------------------------------------------

Unique classes: 3


### HANDLING THE OUTLIERS
QR-based outlier treatment was chosen as it effectively handles extreme values while preserving the underlying data distribution, making it more suitable for model performance compared to winsorization.

In [15]:
# ==== Defining the function for outlier detection and percentage calculation using IQR ====

def detect_outliers(data):
    datas = np.array(data)

    # ==== Quartiles ====
    q1 = np.quantile(datas, 0.25)
    q2 = np.quantile(datas, 0.50)
    q3 = np.quantile(datas, 0.75)

    # ==== IQR & boundsa ====
    IQR = q3 - q1
    lower_bound = q1 - 1.5 * IQR
    upper_bound = q3 + 1.5 * IQR

    # ==== Outlier detection ====
    outliers = datas[(datas < lower_bound) | (datas > upper_bound)]
    outlier_count = len(outliers)
    outlier_percent = round(outlier_count * 100 / len(datas), 2)

    # ==== Display results ====
    print(f"Q1 = {q1}, Q2 (Median) = {q2:.2f}, Q3 = {q3}")
    print(f"IQR = {IQR:.2f}")
    print(f"Lower Bound = {lower_bound:.2f}, Upper Bound = {upper_bound:.2f}")
    print(f"Outliers Detected: {outlier_count}")
    print(f"Outlier Percentage: {outlier_percent}%\n")

    # ==== Calculating IQR, Lower/Upper Bounds, and Outlier Counts for Continuous Numerical Features ====

# Get only numerical columns to apply outlier detection
for feature in data.select_dtypes(include=np.number).columns:
  print(feature,":")
  detect_outliers(data[feature])
  print("-"*50)

Age :
Q1 = 30.0, Q2 (Median) = 36.00, Q3 = 43.0
IQR = 13.00
Lower Bound = 10.50, Upper Bound = 62.50
Outliers Detected: 0
Outlier Percentage: 0.0%

--------------------------------------------------
DistanceFromHome :
Q1 = 2.0, Q2 (Median) = 7.00, Q3 = 14.0
IQR = 12.00
Lower Bound = -16.00, Upper Bound = 32.00
Outliers Detected: 0
Outlier Percentage: 0.0%

--------------------------------------------------
EmpEducationLevel :
Q1 = 2.0, Q2 (Median) = 3.00, Q3 = 4.0
IQR = 2.00
Lower Bound = -1.00, Upper Bound = 7.00
Outliers Detected: 0
Outlier Percentage: 0.0%

--------------------------------------------------
EmpEnvironmentSatisfaction :
Q1 = 2.0, Q2 (Median) = 3.00, Q3 = 4.0
IQR = 2.00
Lower Bound = -1.00, Upper Bound = 7.00
Outliers Detected: 0
Outlier Percentage: 0.0%

--------------------------------------------------
EmpHourlyRate :
Q1 = 48.0, Q2 (Median) = 66.00, Q3 = 83.0
IQR = 35.00
Lower Bound = -4.50, Upper Bound = 135.50
Outliers Detected: 0
Outlier Percentage: 0.0%

------

| **Feature Name**             | **Description**                     | **Outlier %** | **Action Taken** | **Reason**                                           |
| ---------------------------- | ----------------------------------- | ------------- | ---------------- | ---------------------------------------------------- |
| Age                          | Employee age                        | 0.0%          | No action        | Data is well distributed with no extreme values      |
| DistanceFromHome             | Distance between home and office    | 0.0%          | No action        | Values fall within acceptable range                  |
| EmpEducationLevel            | Education level of employee         | 0.0%          | No action        | Discrete categorical-like feature, no outliers       |
| EmpEnvironmentSatisfaction   | Work environment satisfaction       | 0.0%          | No action        | Ordinal scale, no extreme values                     |
| EmpHourlyRate                | Hourly wage of employee             | 0.0%          | No action        | No significant deviation observed                    |
| EmpJobInvolvement            | Level of job involvement            | 0.0%          | No action        | Limited scale (1–4), no outliers                     |
| EmpJobLevel                  | Job seniority level                 | 0.0%          | No action        | Ordinal feature, no anomalies                        |
| EmpJobSatisfaction           | Job satisfaction level              | 0.0%          | No action        | Balanced distribution                                |
| NumCompaniesWorked           | Number of companies worked          | 3.25%         | Kept             | Outliers are realistic (experienced employees)       |
| EmpLastSalaryHikePercent     | Salary hike percentage              | 0.0%          | No action        | No abnormal values                                   |
| EmpRelationshipSatisfaction  | Workplace relationship satisfaction | 0.0%          | No action        | Ordinal scale, no outliers                           |
| TotalWorkExperienceInYears   | Total work experience               | 4.25%         | Kept             | High experience values are valid, not errors         |
| TrainingTimesLastYear        | Number of trainings attended        | 15.67%        | Capped / Kept    | High variation, but meaningful for learning behavior |
| EmpWorkLifeBalance           | Work-life balance rating            | 0.0%          | No action        | Limited scale, no outliers                           |
| ExperienceYearsAtThisCompany | Years at current company            | 4.67%         | Kept             | Long tenure is valid                                 |
| ExperienceYearsInCurrentRole | Years in current role               | 1.33%         | Kept             | Small percentage, acceptable variation               |
| YearsSinceLastPromotion      | Years since last promotion          | 7.33%         | Kept             | Indicates stagnation, important for analysis         |
| YearsWithCurrManager         | Years with current manager          | 0.92%         | No action        | Minimal outliers                                     |
| PerformanceRating            | Employee performance rating         | 27.17%        | No removal       | Target variable; imbalance due to limited categories |


In [16]:
col_to_clip = ['NumCompaniesWorked','TotalWorkExperienceInYears','TrainingTimesLastYear','ExperienceYearsAtThisCompany','ExperienceYearsInCurrentRole', 'YearsSinceLastPromotion','YearsWithCurrManager']

# Keep original data
data_before = data.copy()
data_after = data.copy()

def outliers(data, col):
    Q1 = data[col].quantile(0.25)
    Q3 = data[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    data[col] = data[col].clip(lower, upper)
    return data

# Apply clipping only to the specified columns
for feature_name in col_to_clip:
    data = outliers(data, feature_name)

In [17]:
# =======Remove the irrelvant values========
data= data.iloc[:,1:] # remove the EmpNumber

Irrelevant features are variables that do not contribute meaningful information to the analysis or prediction and can be removed to improve model performance.

> EmpNumber = It Just an ID. So, it will "Remove"

## Feature Engineering

#### Feature Transformation 

#### Temporary Label Encoding for Correlation Analysis

Temporary label encoding was applied to categorical variables solely for the purpose of correlation analysis. This enabled the computation of relationships between features and the target variable (PerformanceRating). 

Based on the correlation results, the top three features most strongly associated with employee performance were identified. This encoding was used only for analysis and was not included in the final model pipeline to avoid introducing bias.

In [18]:
# ===================== for the temporary label encoding  for checking the correlation=======================
from sklearn.preprocessing import LabelEncoder
temp_data = data.copy()
cat_cols=['Gender', 'EducationBackground', 'MaritalStatus',
       'EmpDepartment', 'EmpJobRole', 'BusinessTravelFrequency', 'OverTime',
       'Attrition']
encoder = LabelEncoder()
# Loop through the columns in cat_cols1 and apply LabelEncoder
for col in cat_cols:
    temp_data[col] = encoder.fit_transform(temp_data[col])
temp_data.head(2)

,Age,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,EmpHourlyRate,EmpJobInvolvement,EmpJobLevel,EmpJobSatisfaction,NumCompaniesWorked,OverTime,EmpLastSalaryHikePercent,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,Attrition,PerformanceRating
0,32,1,2,2,5,13,2,10,3,4,55,3,2,4,1.0,0,12,4,10.0,2.0,2,10.0,7.0,0.0,8.0,0,3
1,47,1,2,2,5,13,2,14,4,4,42,3,2,1,2.0,0,12,4,20.0,2.0,3,7.0,7.0,1.0,7.0,0,3


In [19]:
#================Finding out the correlation coeffecient to find out which predictors are significant. =====================
core=temp_data.corr()
core[core>0.8]

,Age,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,EmpHourlyRate,EmpJobInvolvement,EmpJobLevel,EmpJobSatisfaction,NumCompaniesWorked,OverTime,EmpLastSalaryHikePercent,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,Attrition,PerformanceRating
Age,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Gender,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
EducationBackground,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
MaritalStatus,NaN,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
EmpDepartment,NaN,NaN,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
EmpJobRole,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
BusinessTravelFrequency,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
DistanceFromHome,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
EmpEducationLevel,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
EmpEnvironmentSatisfaction,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [20]:
# =========== Identifying Highly Correlated Features ===============
import numpy as np

corr_matrix = temp_data.corr()

# Upper triangle (avoid duplicate pairs)
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))

# Find columns to drop (correlation > 0.8)
to_drop = [column for column in upper.columns if any(upper[column] > 0.8)]

print(to_drop)

['ExperienceYearsInCurrentRole', 'YearsWithCurrManager']


Highly correlated features such as ExperienceYearsInCurrentRole and YearsWithCurrManager were removed to reduce multicollinearity and redundancy, ensuring better model performance and interpretability.

In [21]:
# ========== Create a new feature for control the correlation =================
data['ExperienceRatio'] = data['ExperienceYearsInCurrentRole'] / data['TotalWorkExperienceInYears']

Feature combination means creating a new variable by combining related features to capture meaningful relationships and reduce redundancy.

* **ExperienceRatio tells: How much of total experience is spent in current role.**

In [22]:
# ===========To remove highly correlated (redundant) features ======================
data = data.drop([
    'ExperienceYearsInCurrentRole',
    'YearsWithCurrManager'
], axis=1)

In [23]:
# ============ Check and remove the null values using median ===============
import numpy as np

data['ExperienceRatio'].replace([np.inf, -np.inf], np.nan, inplace=True)
data['ExperienceRatio'].fillna(data['ExperienceRatio'].median(), inplace=True)

Invalid values generated during feature engineering were first handled by converting them to NaN, followed by median imputation to ensure clean and reliable feature representation.

## 2. Top factors effecting performance

In [24]:
# =========== Correlation analysis to find top factors effecting performance ================
# =========== Using based on the temporary label encoding =================
correlation = temp_data.corr()["PerformanceRating"].abs().sort_values(ascending=False)
top_3_factors = correlation[1:4]
print("\nTop 3 Factors effecting Employee Performance:\n", top_3_factors)


Top 3 Factors effecting Employee Performance:
 EmpEnvironmentSatisfaction    0.395561
EmpLastSalaryHikePercent      0.333722
YearsSinceLastPromotion       0.200236
Name: PerformanceRating, dtype: float64


#### Feature Selection Summary:

* There are lot of features in our dataset, to extract important features which really contribute to target feature we need to find corelation coefficient. Corelation coefficient is used to find out the important features in dataset and these features are then used to train the model.
  
* From there, we also get the top factors which affect performance. We can see that the most important features selected were Department, Job Role, Environment Satisfaction, Last Salary Hike Percent, Work Life Balance, Experience Years At This Company, Experience Years In Current Role, Years Since Last Promotion, Years With Current Manager. These were selected because their correlation coeffecient with Performance Rating was more than 0.1.

  
* Standardization and Label Encoding was also used for feature transformation.

  
* **From below corelations we can conclude that top 3 Important Factors affecting employee performance are-**

**Top 3 Factors Effecting Employee Performance**
1. EmpEnvironmentSatisfaction (0.39 correlation) - Employees who rate their work environment highly tend to perform better.
   
2. EmpLastSalaryHikePercent (0.33 correlation) - Higher salary hikes are linked to better performance.
   
4. YearsSinceLastPromotion (0.20 correlation) - Employees who go longer without a promotion tend to perform worse.

In [25]:
import pandas as pd
import os

# Convert array → DataFrame
final_data = pd.DataFrame(data, columns=data.columns)

# Create folder
os.makedirs('data/processed', exist_ok=True)

# Save
final_data.to_csv('data/processed/Cleaned_Data_Emp.csv', index=False)
final_data.to_csv('Cleaned_Data_Emp.csv', index=False)